### DISCUSSION TITLE
Issues in APIRouter.__init__() got an unexpected keyword argument 'prefix' using FastApI

### DISCUSSION THREAD

**Discussion by Nidit2512** [fastapi/fastapi#8552] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from ..import models, schemas
from fastapi import FastAPI, Response, status, HTTPException, Depends, APIRouter 
from sqlalchemy.orm import Session
from .. database import get_db
from typing import List

router = APIRouter(
    prefix="/posts",
    tags=['Posts'])


@router.get("/", response_model=List[schemas.Post])
def get_posts(db: Session = Depends(get_db)):

    posts = db.query(models.Post).all()
    # cursor.execute(""" SELECT * FROM posts """)
    # posts = cursor.fetchall()
    return posts

@router.post("/", status_code= status.HTTP_201_CREATED, response_model=schemas.Post)
def create_posts(post: schemas.PostCreate, db: Session = Depends(get_db)):
    # cursor.execute(""" INSERT INTO posts (title, content, published) VALUES(%s , %s, %s) RETURNING *""",
    # (post.title, post.content, post.published))

    # new_post = cursor.fetchone()
    # conn.commit()
    # print(**post.dict())
    new_post= models.Post(**post.dict())
    db.add(new_post)
    db.commit()
    db.refresh(new_post)
    return new_post

@router.get("/{id}", response_model=schemas.Post)
def get_post(id: int, response: Response, db: Session = Depends(get_db)):
    # cursor.execute(""" SELECT * FROM posts WHERE id = %s """ ,(str(id)))
    # post= cursor.fetchone()
    post = db.query(models.Post).filter(models.Post.id ==id).first()
    print(post)
    if not post:
        raise HTTPException(status_code = status.HTTP_404_NOT_FOUND, 
                                        detail = f"post with id : {id} was not found")
    return post


@router.delete("/{id}", status_code= status.HTTP_204_NO_CONTENT)
def delete_post(id: int, db: Session = Depends(get_db)):
    # cursor.execute (""" DELETE FROM posts WHERE id = %s RETURNING * """, (str(id)))
    # deleted_post= cursor.fetchone()
    # conn.commit()

    post_query = db.query(models.Post).filter(models.Post.id ==id)
    if post_query.first == None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail = f"post with {id} does not exit")

    post_query.delete(synchronize_session=False)
    db.commit()

    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.put("/{id}", response_model=schemas.Post)
def update_post(id: int, updated_post: schemas.PostCreate, db: Session = Depends(get_db)):
    # cursor.execute("""UPDATE posts SET title = %s, content=%s, published=%s WHERE id = %s RETURNING* """, (post.title, post.content, post.published, str(id)),)
    # updated_post = cursor.fetchone()
    # conn.commit()

    post_query = db.query(models.Post).filter(models.Post.id ==id)

    post = post_query.first()
    if post == None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail = f"post with {id} does not exit")
    
    post_query.update(updated_post.dict(), synchronize_session=False)
    db.commit()
    return post_query.first()
```

### Description

 File "C:\Users\OneDrive\Desktop\API_Development\fastapi\.\app\main.py", line 13, in <module>
    from app.routers import post, user
  File "C:\Users\OneDrive\Desktop\API_Development\fastapi\.\app\routers\post.py", line 7, in <module>
    router = APIRouter(
TypeError: APIRouter.__init__() got an unexpected keyword argument 'prefix'

### Operating System

Windows

### FastAPI Version

0.52.0

### Python Version

Python 3.10.1

### Additional Context

Please help

**Comment by rafsaf** [comment]
Your FastAPI version 0.52.0 has not "prefix" argument yet, see

https://github.com/tiangolo/fastapi/blob/025b38df40a18f94322fab445aee13024fe783da/fastapi/routing.py#L341

**Comment by Nidit2512** [comment]
Thank you so much for the response but i wanted to know if i use def __init__ function in my script, would it work?  Or

**Comment by rafsaf** [comment, marked as answer]
Please look at release notes. Your version is kind of 2 years behind. It's 2/3 of whole fastapi lifetime (I'm on phone, this is estimated).

I'm aware of few security issues and breaking changes ;)

For me, upgrade is required.

But, if it's not possible, your only chance is to search through source code of your version and find a solution, probably you can add prefix in `app.include_router`.

**Comment by Nidit2512** [comment]
Hi,

I resolved the above issue after updating the fastapi but I got this new error

(venv) C:\Users\OneDrive\Desktop\API_Development\fastapi>uvicorn app.main:app --reload
INFO:     Will watch for changes in these directories: ['C:\\Users\\OneDrive\\Desktop\\API_Development\\fastapi']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [11924] using watchgod
Process SpawnProcess-1:
Traceback (most recent call last):
  File "C:\Users\AppData\Local\Programs\Python\Python310\lib\multiprocessing\process.py", line 315, in _bootstrap
    self.run()
  File "C:\Users\AppData\Local\Programs\Python\Python310\lib\multiprocessing\process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\OneDrive\Desktop\API_Development\fastapi\venv\lib\site-packages\uvicorn\subprocess.py", line 76, in subprocess_started
    target(sockets=sockets)
  File "C:\Users\OneDrive\Desktop\API_Development\fastapi\venv\lib\site-packages\uvicorn\server.py", line 68, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "C:\Users\AppData\Local\Programs\Python\Python310\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Users\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 641, in run_until_complete
    return future.result()
  File "C:\Users\OneDrive\Desktop\API_Development\fastapi\venv\lib\site-packages\uvicorn\server.py", line 76, in serve
    config.load()
  File "C:\Users\OneDrive\Desktop\API_Development\fastapi\venv\lib\site-packages\uvicorn\config.py", line 448, in load
    self.loaded_app = import_from_string(self.app)
  File "C:\Users\OneDrive\Desktop\API_Development\fastapi\venv\lib\site-packages\uvicorn\importer.py", line 21, in import_from_string
    module = importlib.import_module(module_str)
  File "C:\Users\AppData\Local\Programs\Python\Python310\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Users\OneDrive\Desktop\API_Development\fastapi\.\app\main.py", line 19, in <module>
    app = FASTAPI()
NameError: name 'FASTAPI' is not defined

**Comment by Nidit2512** [comment]
> Hi,
> 
> I resolved the above issue after updating the fastapi but I got this new error
> 
> (venv) C:\Users\OneDrive\Desktop\API_Development\fastapi>uvicorn app.main:app --reload INFO: Will watch for changes in these directories: ['C:\Users\OneDrive\Desktop\API_Development\fastapi'] INFO: Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit) INFO: Started reloader process [11924] using watchgod Process SpawnProcess-1: Traceback (most recent call last): File "C:\Users\AppData\Local\Programs\Python\Python310\lib\multiprocessing\process.py", line 315, in _bootstrap self.run() File "C:\Users\AppData\Local\Programs\Python\Python310\lib\multiprocessing\process.py", line 108, in run self._target(*self._args, **self._kwargs) File "C:\Users\OneDrive\Desktop\API_Development\fastapi\venv\lib\site-packages\uvicorn\subprocess.py", line 76, in subprocess_started target(sockets=sockets) File "C:\Users\OneDrive\Desktop\API_Development\fastapi\venv\lib\site-packages\uvicorn\server.py", line 68, in run return asyncio.run(self.serve(sockets=sockets)) File "C:\Users\AppData\Local\Programs\Python\Python310\lib\asyncio\runners.py", line 44, in run return loop.run_until_complete(main) File "C:\Users\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 641, in run_until_complete return future.result() File "C:\Users\OneDrive\Desktop\API_Development\fastapi\venv\lib\site-packages\uvicorn\server.py", line 76, in serve config.load() File "C:\Users\OneDrive\Desktop\API_Development\fastapi\venv\lib\site-packages\uvicorn\config.py", line 448, in load self.loaded_app = import_from_string(self.app) File "C:\Users\OneDrive\Desktop\API_Development\fastapi\venv\lib\site-packages\uvicorn\importer.py", line 21, in import_from_string module = importlib.import_module(module_str) File "C:\Users\AppData\Local\Programs\Python\Python310\lib\importlib__init__.py", line 126, in import_module return _bootstrap._gcd_import(name[level:], package, level) File "", line 1050, in _gcd_import File "", line 1027, in _find_and_load File "", line 1006, in _find_and_load_unlocked File "", line 688, in _load_unlocked File "", line 883, in exec_module File "", line 241, in _call_with_frames_removed File "C:\Users\OneDrive\Desktop\API_Development\fastapi.\app\main.py", line 19, in app = FASTAPI() NameError: name 'FASTAPI' is not defined

@rafsaf I updated latest fastapi version and after updating I got this error. could you please help me? Thanks

**Comment by adhamsalama** [comment]
@Nidit2512 
it's "FastAPI", not "FASTAPI".
